src/common/gtdb_lineage.py

```diff
@@ -57,19 +66,89 @@ def __ge__(self, other):
     @classmethod
     def from_abbreviation(cls, abbrev: str):
         if not hasattr(cls, "_abbrev_lookup"):
-            cls._abbrev_lookup = {r.rank_abbrev: r for r in cls}
+            cls._abbrev_lookup = {r.abbrev: r for r in cls}
         if abbrev not in cls._abbrev_lookup:
             raise ValueError(f"No such GTDB rank abbreviation: {abbrev}")
         return cls._abbrev_lookup[abbrev]
 
 
-GTDBLineageNode = namedtuple("GTDBLineageNode", "rank name")
+class GTDBLineageNode:
+    """
+    A node in a GTDB lineage.
+
+    Instance variables:
+    rank - the node's rank
+    name - the node's scientific name.
+    """
+    rank: GTDBRank
+    name: str
+
+    def __init__(self, rank: GTDBRank, name: str):
+        """
+        Create the node.
+
+        rank - the node's rank.
+        name - the scientific name of the node.
+        """
+        self.rank = rank
+        self.name = name
+
+    def __str__(self):
+        return f"{self.rank.abbrev}__{self.name}"
+
+    def __eq__(self, other):
+        return (self.rank, self.name) == (other.rank, other.name)
+
+
+class GTDBLineage:
+    """
+    A GTDB lineage represeted as a list of lineage nodes. The lineage may be truncated, but must
+    have all ranks included up to the truncation point.
+
+    Instance variables:
+    lineage - the lineage.
+    """
+
+    lineage: tuple[GTDBLineageNode]
+
+    def __init__(self, lineage: list[GTDBLineageNode]):
+        """
+        Create the lineage.
+        """
+        if not lineage:
+            raise ValueError("lineage required")
+        ranks = [r for r in GTDBRank][:len(lineage)]
+        for n, r in zip(lineage, ranks):
```

Comment:
If the lineage is coming in sorted incorrectly there is something very wrong going on upstream and we should error and fix it rather than silently let it go on being screwed up.

That said, the chance that happens is really low.

Reply:
👍